File: DriftGame/uracer-desktop/src/com/bitfire/uracer/u3d/materials/Material.java
package com.bitfire.uracer.u3d.materials;

import com.badlogic.gdx.graphics.glutils.ShaderProgram;
import com.badlogic.gdx.utils.Array;

import org.jetbrains.annotations.NotNull;

import java.util.Iterator;

public class Material implements Iterable<MaterialAttribute> {
    protected String name;
    /**
     * This flag is true if material contain blendingAttribute
     */
    protected boolean needBlending;
    /**
     * This flag is true if material contain TextureAttribute
     */
    protected boolean hasTexture;
    protected ShaderProgram shader;
    private final Array<MaterialAttribute> attributes;

    public Material(String name, Array<MaterialAttribute> attributes) {
        this.name = name;
        this.attributes = attributes;

        checkAttributes();
    }

    public Material(String name, MaterialAttribute... attributes) {
        this(name, new Array<>(attributes));
    }

    protected void checkAttributes() {
        // this way we foresee if blending is needed with this material and rendering can deferred more easily
        this.needBlending = false;
        this.hasTexture = false;
        for (int i = 0; i < this.attributes.size; i++) {
            if (!needBlending && this.attributes.get(i) instanceof BlendingAttribute)
                this.needBlending = true;
            else if (!hasTexture && this.attributes.get(i) instanceof TextureAttribute) this.hasTexture = true;
        }
    }

    public void bind(ShaderProgram program) {
        for (int i = 0; i < attributes.size; i++) {
            attributes.get(i).bind(program);
        }
    }

    public String getName() {
        return name;
    }

    public Material copy() {
        Array<MaterialAttribute> attributes = new Array<>(this.attributes.size);
        for (int i = 0; i < attributes.size; i++) {
            attributes.add(this.attributes.get(i).copy());
        }
        final Material copy = new Material(name, attributes);
        copy.shader = this.shader;
        return copy;
    }

    @Override
    public int hashCode() {
        final int prime = 31;
        int result = 1;
        result = prime * result + attributes.hashCode();
        result = prime * result + ((name == null) ? 0 : name.hashCode());
        return result;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null) return false;
        if (getClass() != obj.getClass()) return false;
        Material other = (Material) obj;
        if (other.attributes.size != attributes.size) return false;
        for (int i = 0; i < attributes.size; i++) {
            if (!attributes.get(i).equals(other.attributes.get(i))) return false;
        }
        if (name == null) {
            return other.name == null;
        } else return name.equals(other.name);
    }

    public ShaderProgram getShader() {
        return shader;
    }

    public void setShader(final ShaderProgram shader) {
        this.shader = shader;
    }

    @NotNull
    @Override
    public Iterator<MaterialAttribute> iterator() {
        return attributes.iterator();
    }
}

File: DriftGame/uracer-desktop/src/com/bitfire/uracer/u3d/materials/Material.kt
package com.bitfire.uracer.u3d.materials

import com.badlogic.gdx.graphics.glutils.ShaderProgram
import com.badlogic.gdx.utils.Array

class Material(var name: String?, private val attributes: Array<MaterialAttribute>) : Iterable<MaterialAttribute> {

    constructor(name: String, vararg attributes: MaterialAttribute) : this(name, Array<MaterialAttribute>(attributes))

    // This flag is true if the material contains blendingAttribute.
    private var needBlending = false

    // This flag is true if material contains TextureAttribute.
    private var hasTexture = false


    private var shader: ShaderProgram? = null

    init {
        // this way we foresee if blending is needed with this material and rendering can deferred more easily
        for (i in 0..<this.attributes.size) {
            if (!needBlending && this.attributes.get(i) is BlendingAttribute)
                this.needBlending = true
            else if (!hasTexture && this.attributes.get(i) is TextureAttribute)
                this.hasTexture = true
        }
    }

    fun bind(program: ShaderProgram) {
        attributes.forEach { it.bind(program) }
    }

    fun copy(): Material {
        val attributes = Array<MaterialAttribute>(this.attributes.size)
        for (i in 0..<attributes.size) {
            attributes.add(this.attributes.get(i).copy())
        }
        val copy = Material(name, attributes)
        copy.shader = this.shader
        return copy
    }

    override fun hashCode(): Int {
        val prime = 31
        var result = 1
        result = prime * result + attributes.hashCode()
        result = prime * result + (name?.hashCode() ?: 0)
        return result
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null) return false
        if (javaClass != other.javaClass) return false
        val other = other as Material
        if (other.attributes.size != attributes.size) return false
        for (i in 0..<attributes.size) {
            if (attributes.get(i) != other.attributes.get(i)) return false
        }
        return if (name == null) {
            other.name == null
        } else name == other.name
    }

    override fun iterator() = attributes.iterator()

}
